### PR TITLE
276 move evaluate to server

### LINKED ISSUE #276: Move Evaluation to Server
Description:
Move the logic of evaluating an RUL model to the Server package so that the UI won't freeze when a user evaluates a model.

Acceptance Criteria:
[-] UI should not freeze when a user evaluates a model.
[-] UI should automatically update the RMSE value when a new value is in the database.
[-] Evaluating a model should be done on a separate thread.
[-] once an evaluation is done, the new RMSE should be stored in the database.
This issue updates [#216]

### PR DESCRIPTION
**Pull request by TalalBazerbachi**
Evaluating models has been moved to the Server package, whenever a new RMSE value is in the database,  the UI gets updated.
Before you test this branch:
-you should run dump V16 first.
-run server for a few minutes.

This resolves [#276 ]

### CODE REVIEW

File: JavaFXUserInterface/src/main/java/external/ModelDAOImpl.java [hunk 1]
```diff
@@ -172,6 +184,64 @@ public String getGetModelEvaluation(int modelID, String assetTypeID) {
...
         }
-        return rmseValue;
+        if(rmseValue==null) return "n/a";
+        else return rmseValue;
     }
-}
\ No newline at end of file
+
+    @Override
+    public ModelStrategy getModelStrategy(int modelID, int assetTypeID) throws SQLException {
```
Comment (jeffbuonamici) [inline JavaFXUserInterface/src/main/java/external/ModelDAOImpl.java]: Add comments

File: JavaFXUserInterface/src/main/java/external/ModelDAOImpl.java [hunk 2]
```diff
@@ -172,6 +184,64 @@ public String getGetModelEvaluation(int modelID, String assetTypeID) {
...
+                    }}
+            }
+        } catch(SQLException e) {
+            e.printStackTrace();
+        }
+        return modelStrategy;
+    }
+
+    @Override
+    public void updateModelStrategy(ModelStrategy modelStrategy, int modelID, int assetTypeID){
```
Comment (jeffbuonamici) [inline JavaFXUserInterface/src/main/java/external/ModelDAOImpl.java]: Add comments

File: JavaFXUserInterface/src/main/java/external/ModelDAOImpl.java [hunk 3]
```diff
@@ -172,6 +184,64 @@ public String getGetModelEvaluation(int modelID, String assetTypeID) {
...
+            ps.setObject(1, tm.getModelStrategy());
+            ps.setInt(2, tm.getModelID());
+            ps.setInt(3, tm.getAssetTypeID());
+            ps.executeQuery();
+        } catch (SQLException e) {
+            e.printStackTrace();
+        }
+    }
+
+    public double getLatestRMSE(int modelID,int assetTypeID){
```
Comment (jeffbuonamici) [inline JavaFXUserInterface/src/main/java/external/ModelDAOImpl.java]: Add comments

File: Server/src/main/java/app/ModelController.java [hunk 4]
```diff
@@ -248,4 +261,16 @@ public Instances createInstancesFromAssets(List<Asset> assets) {
         }
         return data;
     }
+
+    public void modelToEvaluate(TrainedModel trainedModel) throws Exception {
```
Comment (jeffbuonamici) [inline Server/src/main/java/app/ModelController.java:256]: Add comments

File: Server/src/main/java/app/item/EvaluationList.java [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+//    private EvaluationList(){
+//        modelsList = new ArrayList<EvaluateModel>();
+//    }
+//
+//    private static EvaluationList instance;
+//
+//    public static EvaluationList getInstance() {
+//        if (instance == null) instance = new EvaluationList();
+//        return instance;
+//    }
```
Comment (jeffbuonamici) [inline Server/src/main/java/app/item/EvaluationList.java]: Remove if no longer needed.

File: Server/src/main/java/local/ModelDAO.java [hunk 6]
```diff
@@ -19,6 +19,8 @@
 
     String getModelNameFromModelID(String modelID);
 
+    void updateRMSE(Double rmse, int modelId, int assetTypeId);
+    // TrainedModel createTrainedModelFromResultSet(ResultSet rs, boolean withModel) throws SQLException;
```
Comment (jeffbuonamici) [inline Server/src/main/java/local/ModelDAO.java]: Remove if no longer needed.

File: JavaFXUserInterface/src/main/java/Controllers/AssetTypeInfoController.java [hunk 7]
```diff
@@ -581,41 +528,64 @@ public void generateThumbnails() {
...
+                modelStrategy.setTrainAssets(from);
+                modelStrategy.setTestAssets(to);
+                modelDAO.updateModelStrategy(modelStrategy,1,1);
+                //                    EvaluateModel evaluateModel = new EvaluateModel();
+//                    evaluateModel.setAssetTypeID(assetTypeID);
+//                    evaluateModel.setModelName(modelName);
+//                    evaluateModel.setFrom(from);
+//                    evaluateModel.setTo(to);
+//                    modelList.add(evaluateModel);
+//                    modelDAO.insertToEvaluate(modelName, assetTypeID, (int) trainSlider.getValue(), (int) testSlider.getValue());
```
Comment (jeffbuonamici) [inline JavaFXUserInterface/src/main/java/Controllers/AssetTypeInfoController.java]: Remove if no longer needed.

File: JavaFXUserInterface/src/main/java/Controllers/AssetTypeInfoController.java [hunk 8]
```diff
@@ -581,41 +528,64 @@ public void generateThumbnails() {
...
+                ArrayList<EvaluateModel> modelList = new ArrayList<EvaluateModel>();
+                String modelName = model.getModelName();
+                int assetTypeID = Integer.parseInt(assetType.getId());
+                int from = (int) trainSlider.getValue() + 1;
+                int to = (int) trainSlider.getValue() + 1 + (int) testSlider.getValue();
+                List<Asset> assets= assetDAO.getArchivedAssetsFromAssetTypeID(Integer.parseInt(assetType.getId()));
+                List<Asset> trainAssets= assets.subList(0,from-1);
+                List<Asset> testAssets= assets.subList(from,from+to-1);
+                local.ModelDAOImpl dao=new local.ModelDAOImpl();
+//                    ModelStrategy modelStrategy = modelDAO.getModelStrategy(model.getModelID(),assetTypeID);
```
Comment (jeffbuonamici) [inline JavaFXUserInterface/src/main/java/Controllers/AssetTypeInfoController.java]: Remove if no longer needed.

File: JavaFXUserInterface/src/main/java/Controllers/AssetTypeInfoController.java [hunk 9]
```diff
@@ -284,125 +325,31 @@ public void enableEvaluation(List<Button> enableBtnList) {
...
+//        setTrainAndTestInstances();
+//        ObservableList<Model> selectedModelToEvaluateList = FXCollections.observableArrayList();
+//        selectedModelToEvaluateList.add(selectedModel);
+//        try {
+//            evaluateModelClassifiers(selectedModelToEvaluateList);
+//        } catch (Exception exception) {
+//            exception.printStackTrace();
+//        }
+//        selectedModelToEvaluateList.clear();
+//    }
```
Comment (jeffbuonamici) [inline JavaFXUserInterface/src/main/java/Controllers/AssetTypeInfoController.java]: Remove if no longer needed.

File: JavaFXUserInterface/src/main/java/Controllers/AssetTypeInfoController.java [hunk 10]
```diff
@@ -244,17 +263,39 @@ private void attachEventsModelTab() {
...
+                        } catch (SQLException throwables) {
+                            throwables.printStackTrace();
+                        }
+//                        EvaluateModel evaluateModel = new EvaluateModel();
+//                        evaluateModel.setAssetTypeID(assetTypeID);
+//                        evaluateModel.setModelName(modelName);
+//                        evaluateModel.setFrom(from);
+//                        evaluateModel.setTo(to);
+//                        modelList.add(evaluateModel);
+//                        modelDAO.insertToEvaluate(modelName, assetTypeID, (int) trainSlider.getValue(), (int) testSlider.getValue());
```
Comment (jeffbuonamici) [inline JavaFXUserInterface/src/main/java/Controllers/AssetTypeInfoController.java]: Remove if no longer needed.

File: JavaFXUserInterface/src/main/java/external/ModelDAO.java [hunk 11]
```diff
@@ -8,22 +8,31 @@
...
 public interface ModelDAO {
 
     String getModelNameFromAssetTypeID(String assetTypeId);
 
-    int getModelIDFromAssetTypeID(String assetTypeID);
+    TrainedModel createTrainedModelFromResultSet(ResultSet rs, boolean withModel) throws SQLException;
+    // int getModelIDFromAssetTypeID(String assetTypeID);
 
-    void updateRMSE(Double rmse, int modelId, int assetTypeId);
+    // void updateRMSE(Double rmse, int modelId, int assetTypeId);
```
Comment (jeffbuonamici) [inline JavaFXUserInterface/src/main/java/external/ModelDAO.java]: Remove if no longer needed.

**Review by jeffbuonamici — CHANGES_REQUESTED** [review]

File: JavaFXUserInterface/src/main/java/external/ModelDAOImpl.java (hunk 1 above)
Comment (TalalBazerbachi) [reply JavaFXUserInterface/src/main/java/external/ModelDAOImpl.java]: added comments to all the functions

**Comment by TalalBazerbachi** [comment]
all changes have been applied

File: JavaFXUserInterface/src/main/java/external/ModelDAOImpl.java [hunk 12]
```diff
@@ -19,10 +26,11 @@
 public class ModelDAOImpl extends DAO implements ModelDAO {
     private static final String GET_MODEL_FROM_ASSET_TYPE_ID = "SELECT * from trained_model, model WHERE trained_model.model_id = model.model_id AND trained_model.asset_type_id = ?";
     private static final String GET_ALL_MODELS = "SELECT * from model";
-    private static final String GET_MODEL_EVALUATION = "SELECT rmse FROM model_evaluation WHERE model_id = ? AND asset_type_id = ?";
-    private static final String INSERT_RMSE = "REPLACE INTO model_evaluation SET rmse = ?,model_id = ?, asset_type_id = ? ";
+    private static final String GET_LATEST_RMSE = "select rmse from trained_model where model_id=? and asset_type_id=? and status_id=2";
     private static final String UPDATE_MODEL_FOR_ASSET_TYPE = "UPDATE trained_model set model_id = ? where asset_type_id = ?";
     private static final String UPDATE_RETRAIN = "UPDATE trained_model SET retrain = true WHERE asset_type_id = ?";
+    private static final String GET_MODEL_STRATEGY = "select serialized_model from trained_model where model_id=? and asset_type_id=? and status_id=2";
```
Comment (najimgh) [inline JavaFXUserInterface/src/main/java/external/ModelDAOImpl.java]: It would be good to have SQL keywords in caps as it is our convention

File: JavaFXUserInterface/src/main/java/external/ModelDAOImpl.java [hunk 13]
```diff
@@ -19,10 +26,11 @@
 public class ModelDAOImpl extends DAO implements ModelDAO {
     private static final String GET_MODEL_FROM_ASSET_TYPE_ID = "SELECT * from trained_model, model WHERE trained_model.model_id = model.model_id AND trained_model.asset_type_id = ?";
     private static final String GET_ALL_MODELS = "SELECT * from model";
-    private static final String GET_MODEL_EVALUATION = "SELECT rmse FROM model_evaluation WHERE model_id = ? AND asset_type_id = ?";
-    private static final String INSERT_RMSE = "REPLACE INTO model_evaluation SET rmse = ?,model_id = ?, asset_type_id = ? ";
+    private static final String GET_LATEST_RMSE = "select rmse from trained_model where model_id=? and asset_type_id=? and status_id=2";
```
Comment (najimgh) [inline JavaFXUserInterface/src/main/java/external/ModelDAOImpl.java]: It would be good to have SQL keywords in caps as it is our convention

**Review by najimgh — CHANGES_REQUESTED** [review]
You should add your changes to the database on top of the most recent cbms_script version (dumpfile v15 ?)

File: JavaFXUserInterface/src/main/java/external/ModelDAOImpl.java (hunk 13 above)
Comment (TalalBazerbachi) [reply JavaFXUserInterface/src/main/java/external/ModelDAOImpl.java]: @najimgh I updated the queries

**Comment by TalalBazerbachi** [comment]
@najimgh yes after you run cbms script 15, run script 16 (NOTE: if you ran v15 before, run it again)

**Comment by rdsaliba** [comment]
Hi @t98b. Everything seems fine to me. There's still the commented code as mentioned by Jeff thought.

**Comment by TalalBazerbachi** [comment]
@rdsaliba 
can you please provide screenshots, so I know where there is still commented code?
Because I did remove all commented code. 
I might have missed some

**Comment by rdsaliba** [comment]
image

image

**Comment by rdsaliba** [comment]
There's more. Look at Jeff comment's and resolve those. They indicate where in the code.

**Comment by TalalBazerbachi** [comment]
removed them

File: Server/src/main/java/app/item/EvaluationList.java (hunk 5 above)
Comment (rdsaliba) [reply Server/src/main/java/app/item/EvaluationList.java]: @t98b You still have this one!

File: Server/src/main/java/app/ModelController.java (hunk 4 above)
Comment (rdsaliba) [reply Server/src/main/java/app/ModelController.java:256]: @t98b There's multiple review by Jeff asking for comments. Did you add those? Look at the comments and resolve the conversations.

**Comment by TalalBazerbachi** [comment]
I fixed all the changes requested

File: JavaFXUserInterface/src/main/java/external/ModelDAO.java [hunk 14]
```diff
@@ -8,22 +8,28 @@
...
 import java.util.List;
 
 public interface ModelDAO {
 
     String getModelNameFromAssetTypeID(String assetTypeId);
 
-    int getModelIDFromAssetTypeID(String assetTypeID);
-
-    void updateRMSE(Double rmse, int modelId, int assetTypeId);
+    TrainedModel createTrainedModelFromResultSet(ResultSet rs, boolean withModel) throws SQLException;
```
Comment (rdsaliba) [inline JavaFXUserInterface/src/main/java/external/ModelDAO.java:22]: This method is never used. Why did you add it?

Comment (t98b) [reply JavaFXUserInterface/src/main/java/external/ModelDAO.java:22]: getModelEvaluation?

**Review by Khal-Shah — CHANGES_REQUESTED** [review]
I ran V16 after V15 and ran server for several minutes but keep getting these errors: 
image

Also, the ZenHub for this issue is blank, could you put a brief description and acceptance criteria there for this issue and also reference the previous issue this fixes.

**Comment by PaulMicu** [comment]
> I ran V16 after V15 and ran server for several minutes but keep getting these errors:
> image
> 
> Also, the ZenHub for this issue is blank, could you put a brief description and acceptance criteria there for this issue and also reference the previous issue this fixes.

this is a local issue, you need to go increase the max_allowed_packet on your local database

File: Server/src/main/java/app/ModelController.java [hunk 15]
```diff
@@ -124,12 +123,31 @@ public void checkModels() {
...
                 .forEach(this::trainAndSave);
+        trainedModelsToRetrain
+                .stream()
+                .filter(tm -> tm.getStatusID() == Constants.STATUS_EVALUATION)
+                .findFirst().ifPresent(this::evaluate);
         trainedModelsToRetrain
                 .stream()
                 .filter(tm -> tm.getStatusID() == Constants.STATUS_EVALUATION)
                 .findFirst().ifPresent(this::trainAndSave);
     }
```
Comment (PaulMicu) [inline Server/src/main/java/app/ModelController.java]: you cant do that....like i told you, simply add an if statement in trainAndSave after the model is trained to check if the statusID is Eval and then do the evaluate

File: Server/src/main/java/app/ModelController.java [hunk 16]
```diff
@@ -124,12 +123,31 @@ public void checkModels() {
...
+     */
+    public void evaluate(TrainedModel tm){
+        try{
+            if(!Objects.isNull(tm.getModelStrategy()));
+            modelToEvaluate(tm);
+        }
+        catch(Exception e){
+            e.printStackTrace();
+        }
+    }
```
Comment (PaulMicu) [inline Server/src/main/java/app/ModelController.java]: you dont need this

File: Server/src/main/java/app/ModelController.java [hunk 17]
```diff
@@ -249,4 +265,21 @@ public Instances createInstancesFromAssets(List<Asset> assets) {
...
+
+    /**
+     * evaluates a model on a separate thread using the evaluation class
+     * @param trainedModel
+     * @author talal
+     */
+    public void modelToEvaluate(TrainedModel trainedModel) throws Exception {
+        trainedModel.setAssetTypeID(1);
+        int trainAssets =trainedModel.getModelStrategy().getTrainsAssets();
+        int testAssets=trainedModel.getModelStrategy().getTestAssets();
```
Comment (PaulMicu) [inline Server/src/main/java/app/ModelController.java]: its better to keep track of an ArrayList of assets instead of an int to avoid calling assetDaoImpl.getAssetsFromAssetTypeID(trainedModel.getAssetTypeID()) multiple times for no reason

File: Server/src/main/java/app/item/Model.java [hunk 18]
```diff
@@ -1,12 +1,20 @@
...
     private String modelName;
     private int modelID;
     private String description;
+    private SimpleStringProperty rmse;
+
+    public Model() { rmse = new SimpleStringProperty();}
+
+    public SimpleStringProperty getRMSE(){ return rmse;}
+
+    public void setRMSE(String rmse){ this.rmse.setValue(rmse);}
```
Comment (PaulMicu) [inline Server/src/main/java/app/item/Model.java]: you need to use TrainedModel to store the RMSE not Model

File: Server/src/main/java/app/ModelController.java [hunk 19]
```diff
@@ -219,21 +237,19 @@ public Double estimateRUL(Asset asset, Classifier classifier) {
      * @author Paul
      */
     public Instances createInstancesFromAssets(List<Asset> assets) {
+        System.out.println("in create");
```
Comment (PaulMicu) [inline Server/src/main/java/app/ModelController.java]: remove this

File: JavaFXUserInterface/src/main/java/external/ModelDAOImpl.java [hunk 20]
```diff
@@ -106,6 +116,13 @@ public void updateRMSE(Double rmse, int modelId, int assetTypeId) {
                     newModel.setModelID(rs.getInt("model_id"));
                     newModel.setModelName(rs.getString("name"));
                     newModel.setDescription(rs.getString("description"));
+                    try (PreparedStatement pss = getConnection().prepareStatement(GET_LATEST_RMSE)) {
+                        pss.setInt(1, newModel.getModelID());
+                        pss.setString(2, String.valueOf(assetTypeID));
+                        try (ResultSet rss = pss.executeQuery()) {
+                            if (rss.next())
+                                newModel.setRMSE(rss.getString("rmse"));
+                        }}
```
Comment (PaulMicu) [inline JavaFXUserInterface/src/main/java/external/ModelDAOImpl.java]: Avoid doing 1 query per model, simply get the trainedmodels with status id 2 for the specified asset type id and then iterate over that list.. you will then only need 1 query instead of 9...

**Review by PaulMicu — CHANGES_REQUESTED** [review]

File: Server/src/main/java/app/ModelController.java (hunk 17 above)
Comment (TalalBazerbachi) [reply Server/src/main/java/app/ModelController.java]: it doesn't work, you can't have a list in model strategy, it caused a nonTransient error, I tried to fix it for hours but it didn't work

File: Server/src/main/java/app/ModelController.java (hunk 16 above)
Comment (TalalBazerbachi) [reply Server/src/main/java/app/ModelController.java]: function removed

File: Server/src/main/java/app/ModelController.java (hunk 15 above)
Comment (TalalBazerbachi) [reply Server/src/main/java/app/ModelController.java]: removed this code

File: Server/src/main/java/app/item/Model.java (hunk 18 above)
Comment (TalalBazerbachi) [reply Server/src/main/java/app/item/Model.java]: I am using Model because Jeremie used Model to do his issue last sprint, I just added on top of what he did

File: Server/src/main/java/app/ModelController.java (hunk 19 above)
Comment (TalalBazerbachi) [reply Server/src/main/java/app/ModelController.java]: removed print statement

File: JavaFXUserInterface/src/main/java/controllers/AssetTypeInfoController.java [hunk 21]
```diff
@@ -249,14 +264,33 @@ private void attachEventsModelTab() {
...
+                            ModelStrategy modelStrategy = modelDAO.getModelStrategy(associatedModelID,assetTypeID);
+                            modelStrategy.setTrainAssets(trainAssets);
+                            modelStrategy.setTestAssets(testAssets);
+                            modelDAO.updateModelStrategy(modelStrategy,associatedModelID,assetTypeID);
+                        } catch (SQLException throwables) {
+                            throwables.printStackTrace();
+                        }
+                         }
+
+                });
```
Comment (PaulMicu) [inline JavaFXUserInterface/src/main/java/controllers/AssetTypeInfoController.java]: This code is duplicated almost line by line on line 510, please reduce the duplication

**Review by PaulMicu — DISMISSED** [review]

File: JavaFXUserInterface/src/main/java/external/ModelDAOImpl.java (hunk 20 above)
Comment (TalalBazerbachi) [reply JavaFXUserInterface/src/main/java/external/ModelDAOImpl.java]: changes applied

File: JavaFXUserInterface/src/main/java/controllers/AssetTypeInfoController.java (hunk 21 above)
Comment (TalalBazerbachi) [reply JavaFXUserInterface/src/main/java/controllers/AssetTypeInfoController.java]: moved duplicated code to separate function

**Review by Khal-Shah — DISMISSED** [review]

**Comment by TalalBazerbachi** [comment]
@PaulMicu unused objects and imports have been removed

File: Server/src/main/java/app/item/Model.java (hunk 18 above)
Comment (JayChouteau) [reply Server/src/main/java/app/item/Model.java]: The RMSE is now stored in the trained model table

File: Server/src/main/java/app/ModelController.java (hunk 17 above)
Comment (JayChouteau) [reply Server/src/main/java/app/ModelController.java]: I believe we fixed the transient error for the most part. However, the changing of int to an arraylist of assets will be postpone to a further issue due to time constraints.

**Review by JayChouteau — DISMISSED** [review]
Good work on moving the evaluation to server. It seems like the running the server and evaluating now works as intended. However, you need to let the server run for a while to train all models at least once in the first place.

Another comment, please make sure to run the code-clean-up, address warnings when you push code. Also, please be sure to comment where needed and give meaningful names to variables and functions when you create or edit them so that someone can understand what the code does by just reading the names.

**Review by jeffbuonamici — DISMISSED** [review]

**Review by jeffbuonamici — APPROVED** [review]

**Review by najimgh — APPROVED** [review]

**Comment by Khal-Shah** [comment]
The process for updating the RMSE is not very efficient or user friendly because of the decoupling of UI and server and how the a model has to be trained and saved in order for the RMSE to get saved to the DB. But the feature works as intended if the steps of having the serializable object in the DB and then running server after pressing on evaluate are followed correctly. Initially the RMSE was not aligned well with the evaluate button but @JayChouteau fixed that bug within this issue. Very good addition with the alert box as well. 

Approved.

**Review by Khal-Shah — APPROVED** [review]
